crates/sui-move-build/src/lib.rs

```diff
@@ -66,6 +67,9 @@ pub struct CompiledPackage {
     pub dependency_ids: PackageDependencies,
     /// Path to the Move package (i.e., where the Move.toml file is)
     pub path: PathBuf,
+    /// The bytecode modules that this package depends on (both directly and transitively),
+    /// i.e. on-chain dependencies.
+    pub bytecode_deps: Vec<(PackageName, CompiledModule)>,
```

Comment:
Should we move this inside `MoveCompiledPackage` instead?

Reply:
Yeah this definitely feels like something that should be in the core side rather than the sui side

Reply:
Agree. It could be added to `MoveCompiledPackage` in `fn build_all` of `external-crates/move/crates/move-package/src/compilation/compiled_package.rs`

I'll make a concrete proposal with slightly more details here and get your thoughts @kklas: what if we included on-chain dependencies in the existing `deps_compiled_units` of `MoveCompiledPackage` instead of a separate `bytecode_deps`? I.e., we can decode / deserialize in `build_all` and append to `bytecode_deps`? I think I tried something like this a while back, for keeping compiled artifacts, for a separate reason.

It might need some chances to [deps_compiled_units](https://sourcegraph.com/github.com/MystenLabs/sui@1b7244a274353d619cb67b33ee397c41de3528ea/-/blob/external-crates/move/crates/move-package/src/compilation/compiled_package.rs?L84) where where we have `CompiledUnitWithSource` (either turning the bytecode deps into a `CompiledUnitWithSource`, or creating an enum here for `CompiledBytecode`).

That feels a little less "tacked-on" than a separate vec. It would have some follow on effect on https://github.com/MystenLabs/sui/pull/16523/commits/3c976f2f2da7aededf792bc1ad067fa24f87f67a when we read `bytecode_deps` for source verification to skip on-chain bytecode, and maybe it's a bit more elegant to have all compiled dependencies in one vec (including on-chain ones)?

If that seems too tricky, having a separate vec for `bytecode_deps` in `MoveCompiledPackage` is probably an okay tradeoff (?)

Reply:
> what if we included on-chain dependencies in the existing deps_compiled_units of MoveCompiledPackage instead of a separate bytecode_deps

To me it feels a bit strange to put it in "deps_compiled_units" since these were technically not compiled. I'd rather keep it separate and then just include them in the helper methods on `MoveCompiledPacakge`. Perhaps @tnowacki has some thoughts.

Reply:
In a sense, they were pre-compiled in the past :-) But it's fair to want to distinguish between compiled units of source and those pulled from on-chain. So I'll let others weigh in, just mentioning that possibility!

Reply:
Makes sense and also see my comment here https://github.com/MystenLabs/sui/pull/16523#discussion_r1740558037. I'm not nearly as familiar with this codebase as any of you are and I'm also not the one going to have live with this code, so LMK what you want me to do here. I open to both.